File: python-project-euler-100/p020.py
import time
import math
from Solution import Solution

class P020(Solution):
    mem_cache = {0: 1}
        
    def digits_sum(self, segment):
        sum = 0
        for i in segment:
            sum += int(i)
        return sum

    def factorial(self, n):
        if n in self.mem_cache:
            return self.mem_cache[n]
        data = self.factorial(n - 1)
        self.mem_cache[n] = data * n
        return self.mem_cache[n] 

    def solve(self):
        self.problem_number = 20
        n = 100

        for i in range(1, n):
            self.factorial(i)

        result = self.digits_sum(str(self.factorial(n)))

        return result

def main():
    P020().run()

if __name__ == "__main__":
    main()

File: go-project-euler-100/p020.go
package main

import "math/big"

var p020Cache = map[int]*big.Int{0: big.NewInt(1)}

func p020Factorial(n int) *big.Int {
	if v, ok := p020Cache[n]; ok {
		return v
	}
	data := new(big.Int).Mul(big.NewInt(int64(n)), p020Factorial(n-1))
	p020Cache[n] = data
	return data
}

/*P020Solve solution for p020*/
func P020Solve() int {
	n := 100
	result := 0
	for _, ch := range p020Factorial(n).String() {
		result += int(ch - '0')
	}
	return result
}
